File: app/src/main/java/com/airhud/app/data/HudRepository.kt
package com.airhud.app.data

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * Process-wide singleton that mirrors the "HUD Data Layer / State Stream" from the spec.
 * The overlay service, notification listener, battery receiver and settings UI all
 * read/write through this single StateFlow instead of a real IPC bus, since everything
 * runs in the same app process on this Android-only MVP.
 */
object HudRepository {

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private var notificationClearJob: Job? = null
    private var aiClearJob: Job? = null

    private val _state = MutableStateFlow(HudRuntimeState())
    val state: StateFlow<HudRuntimeState> = _state

    fun setServiceRunning(running: Boolean) {
        _state.update { it.copy(serviceRunning = running) }
    }

    fun updateClock(clockText: String, dateText: String) {
        _state.update { it.copy(clockText = clockText, dateText = dateText) }
    }

    fun updateBattery(percent: Int, isCharging: Boolean) {
        _state.update { it.copy(batteryPercent = percent, isCharging = isCharging) }
    }

    fun postNotification(info: NotificationInfo, autoClearMillis: Long = 5000L) {
        notificationClearJob?.cancel()
        _state.update { it.copy(notification = info) }
        notificationClearJob = scope.launch {
            delay(autoClearMillis)
            _state.update { current ->
                if (current.notification?.timestampMillis == info.timestampMillis) {
                    current.copy(notification = null)
                } else {
                    current
                }
            }
        }
    }

    fun setAiPrompt(prompt: String) {
        _state.update { it.copy(aiPrompt = prompt) }
    }

    fun setAiLoading() {
        aiClearJob?.cancel()
        _state.update { it.copy(aiLoading = true, aiError = null) }
    }

    fun setAiResult(response: String, autoClearMillis: Long = 20_000L) {
        _state.update { it.copy(aiLoading = false, aiResponse = response, aiError = null) }
        aiClearJob?.cancel()
        aiClearJob = scope.launch {
            delay(autoClearMillis)
            _state.update { it.copy(aiResponse = null) }
        }
    }

    fun setAiError(message: String) {
        _state.update { it.copy(aiLoading = false, aiError = message) }
    }

    fun dismissNotification() {
        notificationClearJob?.cancel()
        _state.update { it.copy(notification = null) }
    }

    fun dismissAiResponse() {
        aiClearJob?.cancel()
        _state.update { it.copy(aiResponse = null, aiError = null) }
    }
}

File: unity/AirHUD-Spatial/Assets/Scripts/Data/HudRepository.cs
using System;
using System.Collections;
using UnityEngine;

namespace AirHud.Data
{
    /// <summary>
    /// Runtime "HUD Data Layer / State Stream" from spec section 6. Every widget
    /// reads from this single state object and reacts to <see cref="StateChanged"/>;
    /// on the phone side (Kotlin) the analogous type is HudRepository/StateFlow.
    /// Once the Android app is bridged to this Unity renderer, its data should be
    /// pushed in here instead of the mock producers under Widgets/.
    /// </summary>
    public static class HudRepository
    {
        public static HudRuntimeState State { get; } = new HudRuntimeState();
        public static event Action<HudRuntimeState> StateChanged;

        private static Runner runner;
        private static Coroutine notificationClearRoutine;
        private static Coroutine aiClearRoutine;

        [RuntimeInitializeOnLoadMethod(RuntimeInitializeLoadType.BeforeSceneLoad)]
        private static void EnsureRunner()
        {
            if (runner != null) return;
            var go = new GameObject("HudRepositoryRunner");
            UnityEngine.Object.DontDestroyOnLoad(go);
            runner = go.AddComponent<Runner>();
        }

        private static void Notify() => StateChanged?.Invoke(State);

        public static void SetServiceRunning(bool running)
        {
            State.serviceRunning = running;
            Notify();
        }

        public static void UpdateClock(string clockText, string dateText)
        {
            State.clockText = clockText;
            State.dateText = dateText;
            Notify();
        }

        public static void UpdateBattery(int percent, bool isCharging)
        {
            State.batteryPercent = percent;
            State.isCharging = isCharging;
            Notify();
        }

        public static void PostNotification(NotificationInfo info, float autoClearSeconds = 5f)
        {
            EnsureRunner();
            if (notificationClearRoutine != null) runner.StopCoroutine(notificationClearRoutine);
            State.notification = info;
            Notify();
            notificationClearRoutine = runner.StartCoroutine(ClearNotificationAfter(info, autoClearSeconds));
        }

        private static IEnumerator ClearNotificationAfter(NotificationInfo info, float seconds)
        {
            yield return new WaitForSeconds(seconds);
            if (State.notification == info)
            {
                State.notification = null;
                Notify();
            }
        }

        public static void SetAiPrompt(string prompt)
        {
            State.aiPrompt = prompt;
            Notify();
        }

        public static void SetAiLoading()
        {
            EnsureRunner();
            if (aiClearRoutine != null) runner.StopCoroutine(aiClearRoutine);
            State.aiLoading = true;
            State.aiError = null;
            Notify();
        }

        public static void SetAiResult(string response, float autoClearSeconds = 20f)
        {
            EnsureRunner();
            State.aiLoading = false;
            State.aiResponse = response;
            State.aiError = null;
            Notify();
            if (aiClearRoutine != null) runner.StopCoroutine(aiClearRoutine);
            aiClearRoutine = runner.StartCoroutine(ClearAiAfter(seconds: autoClearSeconds));
        }

        public static void SetAiError(string message)
        {
            State.aiLoading = false;
            State.aiError = message;
            Notify();
        }

        private static IEnumerator ClearAiAfter(float seconds)
        {
            yield return new WaitForSeconds(seconds);
            State.aiResponse = null;
            Notify();
        }

        private class Runner : MonoBehaviour { }
    }
}
